checker.py

```diff
@@ -232,9 +232,20 @@ def _test_file_fragment(self, fname, fragment):
             if not os.path.exists(fname):
                 return False
 
-            with open(fname) as fdata:
-                soup = BeautifulSoup(fdata, "html.parser")
-                self.soup_mapping[fname] = soup
+            if os.path.isfile(fname):
+                filename = fname
+            elif os.path.isdir(fname):
+                filename = False
+                while not filename:
+                    for possiblefile in ['index.htm', 'index.html']:
+                        if os.path.isfile(os.path.join(fname, possiblefile)):
+                            filename = os.path.join(fname, possiblefile)
+            if filename:
```

Comment:
this is a bit verbose. will something like this do the trick for your usecase?:
```diff
+def first_existing_candidate(pth):
+    """return first existing file URL for given directory URL or None"""
+    candidates = [os.path.join(pth, p) for p in ("index.html", "index.htm")]
+    return next(filter(os.path.isfile, [pth] + candidates), None)
```

```diff
    def _test_file_fragment(self, fname, fragment):
         if fname in self.soup_mapping:
             soup = self.soup_mapping[fname]
         else:
-            if not os.path.exists(fname):
+            fname = first_existing_candidate(fname)
+            if fname is None:
                 return False
```

Reply:
Hi, that looks a good approach, so go ahead for it.

About my pelican case:

- Pelican does render markdown into articles and categories, etc for creating a blog, the 'root' for categories or tag is linked as /categories or /tags with no 'file'.

In my case, when applying the test against my local file system (output/ folder, where Pelican generates the files), that folder is there, but serves no file (on an http server it will default to server's 'index.htm, html, default.htm, etc)


So for a reproducer, create a file linking to a folder with an index.htm file and have the first file just link to the folder, without adding the filename

Does it help?

Thanks!
Pablo